File: backend/app/schemas/tutor.py
import uuid
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field
from app.models.tutor import TutorStatus, TutorDifficulty, FlashcardRating


class TopicScope(BaseModel):
    type: Literal["ENTIRE_DOCUMENT", "CHAPTER", "SECTION", "SELECTION", "HIGHLIGHTS"] = "ENTIRE_DOCUMENT"
    target: Optional[str] = None
    page_number: Optional[int] = None
    selected_text: Optional[str] = None


class CreateStudySessionRequest(BaseModel):
    document_id: uuid.UUID
    title: Optional[str] = None
    topic_scope: Optional[TopicScope] = None
    difficulty: TutorDifficulty = TutorDifficulty.INTERMEDIATE
    learning_goal: str = "UNDERSTAND"


class StudySessionResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    document_id: uuid.UUID
    title: str
    topic_scope: Dict[str, Any]
    difficulty: TutorDifficulty
    learning_goal: str
    status: TutorStatus
    current_concept: Optional[str] = None
    mastery_state: Dict[str, Any] = Field(default_factory=dict)
    progress_pct: float = 0.0
    history: List[Dict[str, Any]] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TutorTurnRequest(BaseModel):
    action: Optional[Literal["START", "ANSWER", "NEXT_CONCEPT", "RETRY"]] = "ANSWER"
    user_answer: Optional[str] = None


class TutorHintResponse(BaseModel):
    session_id: uuid.UUID
    hint_level: int  # 1, 2, 3
    hint_text: str


class StudySummaryResponse(BaseModel):
    session_id: uuid.UUID
    document_id: uuid.UUID
    summary_text: str
    concepts_mastered: List[str]
    concepts_needing_review: List[str]
    overall_performance_score: float
    created_at: datetime


class GenerateFlashcardsRequest(BaseModel):
    document_id: uuid.UUID
    session_id: Optional[uuid.UUID] = None
    topic_scope: Optional[TopicScope] = None
    count: int = Field(default=5, ge=1, le=20)


class FlashcardResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    document_id: uuid.UUID
    session_id: Optional[uuid.UUID] = None
    front: str
    back: str
    source_citations: List[Dict[str, Any]] = Field(default_factory=list)
    difficulty: TutorDifficulty
    review_rating: Optional[FlashcardRating] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class RateFlashcardRequest(BaseModel):
    rating: FlashcardRating


class GenerateQuizRequest(BaseModel):
    document_id: uuid.UUID
    session_id: Optional[uuid.UUID] = None
    topic_scope: Optional[TopicScope] = None
    question_count: int = Field(default=5, ge=1, le=10)


class SubmitQuizRequest(BaseModel):
    answers: Dict[str, str]  # question_id -> user choice or text answer


class QuizResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    document_id: uuid.UUID
    session_id: Optional[uuid.UUID] = None
    title: str
    topic_scope: Dict[str, Any]
    questions: List[Dict[str, Any]] = Field(default_factory=list)
    user_answers: Dict[str, Any] = Field(default_factory=dict)
    evaluation_results: Dict[str, Any] = Field(default_factory=dict)
    completed_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/tutor.ts
export type TutorDifficulty = 'BEGINNER' | 'INTERMEDIATE' | 'ADVANCED';

export type TutorStatus =
  | 'CREATED'
  | 'ASSESSING'
  | 'TEACHING'
  | 'QUESTION'
  | 'EVALUATING'
  | 'FEEDBACK'
  | 'NEXT_CONCEPT'
  | 'COMPLETED';

export type FlashcardRating = 'AGAIN' | 'HARD' | 'GOOD' | 'EASY';

export interface TopicScope {
  type: 'ENTIRE_DOCUMENT' | 'CHAPTER' | 'SECTION' | 'SELECTION' | 'HIGHLIGHTS';
  target?: string;
  page_number?: number;
  selected_text?: string;
}

export interface CreateStudySessionRequest {
  document_id: string;
  title?: string;
  topic_scope?: TopicScope;
  difficulty?: TutorDifficulty;
  learning_goal?: string;
}

export interface StudySession {
  id: string;
  user_id: string;
  document_id: string;
  title: string;
  topic_scope: TopicScope;
  difficulty: TutorDifficulty;
  learning_goal: string;
  status: TutorStatus;
  current_concept?: string;
  mastery_state: Record<string, 'NOT_STARTED' | 'NEEDS_REVIEW' | 'MOSTLY_UNDERSTOOD' | 'MASTERED'>;
  progress_pct: number;
  history: Array<{
    turn_id: string;
    state: TutorStatus;
    concept?: string;
    user_answer?: string;
    diagnostic_answer?: string;
    tutor_message?: string;
    question?: string;
    score?: number;
    feedback?: string;
    hints?: string[];
    hints_used?: number;
    created_at: string;
  }>;
  created_at: string;
  updated_at: string;
}

export interface TutorTurnRequest {
  action?: 'START' | 'ANSWER' | 'NEXT_CONCEPT' | 'RETRY';
  user_answer?: string;
}

export interface TutorHintResponse {
  session_id: string;
  hint_level: number;
  hint_text: string;
}

export interface StudySummary {
  session_id: string;
  document_id: string;
  summary_text: string;
  concepts_mastered: string[];
  concepts_needing_review: string[];
  overall_performance_score: number;
  created_at: string;
}

export interface GenerateFlashcardsRequest {
  document_id: string;
  session_id?: string;
  topic_scope?: TopicScope;
  count?: number;
}

export interface Flashcard {
  id: string;
  user_id: string;
  document_id: string;
  session_id?: string;
  front: string;
  back: string;
  source_citations: Array<{
    chunk_id?: string;
    page_number?: number;
    snippet?: string;
  }>;
  difficulty: TutorDifficulty;
  review_rating?: FlashcardRating;
  created_at: string;
}

export interface RateFlashcardRequest {
  rating: FlashcardRating;
}

export interface GenerateQuizRequest {
  document_id: string;
  session_id?: string;
  topic_scope?: TopicScope;
  question_count?: number;
}

export interface QuizQuestion {
  id: string;
  type: 'MCQ' | 'TRUE_FALSE' | 'SHORT_ANSWER';
  question: string;
  options?: string[];
  correct_answer: string;
  explanation: string;
  citation_page?: number;
}

export interface Quiz {
  id: string;
  user_id: string;
  document_id: string;
  session_id?: string;
  title: string;
  topic_scope: TopicScope;
  questions: QuizQuestion[];
  user_answers: Record<string, string>;
  evaluation_results: Record<
    string,
    {
      user_answer: string;
      correct_answer: string;
      is_correct: boolean;
      score: number;
      explanation: string;
    }
  >;
  completed_at?: string;
  created_at: string;
}

export interface SubmitQuizRequest {
  answers: Record<string, string>;
}
